Yoast/duplicate-post
# Options_Form_Generator::sort_taxonomy_objects(): improve sort order and fix PHP 8 compatibility

## Context

* Improve PHP 8 compatibility

## Summary

This PR can be summarized in the following changelog entry:

* Improved sort order of the taxonomies list on the options page.
* Improved PHP 8.0 compatibility.

## Relevant technical choices:

👉 Recommended to review commit by commit.

### Tests: expand the test for Options_Form_Generator::generate_taxonomy_exclusion_list()

The `Options_Form_Generator::generate_taxonomy_exclusion_list()` function uses a custom sort function, but that function only sorts on public versus private, not on name, as demonstrated by this expanded test.

### Options_Form_Generator::sort_taxonomy_objects(): improve sort order and fix PHP 8 compatibility

The `u*sort()` functions in PHP have always been expected to return an integer: a negative integer for when A should be considered "less than" B, a `0` when they should be considered as equal and positive integer for when B should be considered "more than" A.

The `Options_Form_Generator::sort_taxonomy_objects()` did not comply with this and returned a boolean value instead.

As of PHP 8.0, PHP checks more strictly on the callback implementation being correct and will throw a `usort(): Returning bool from comparison function is deprecated, return an integer less than, equal to, or greater than zero` warning.

This warning has now been mitigated by a slightly changed implementation, which will return an integer, as expected.

Additionally, as highlighted by the test expansion in the previous commit, the sort order for when the visibility of two taxonomies was considered "equal" was previously undefined.

This commit also fixes that and will now sort two taxonomies with the same visibility, case-insensitively by their public label using natural sort order.

The unit test has been adjusted to reflect the improve sorting.

Refs:
* https://wiki.php.net/rfc/stable_sorting
* https://www.php.net/manual/en/function.usort.php
* https://www.php.net/manual/en/function.strnatcasecmp.php



## Test instructions

### Test instructions for the acceptance test before the PR gets merged
This PR can be acceptance tested by following these steps:

Preparation:
* Make sure there are at least four taxonomies registered with the test install, some `public`, some `private`.

Testing:
* Check out `trunk`.
* Go to the admin page and observe the sort order of the taxonomy exclusion list.
    This list will be sorted by "public first, private second", but without a fixed sort order within the "public" or "private" taxonomies. Most likely it will sort based on the order of taxonomy registration with WP.
* Switch to PHP 8.0.
* Reload the admin page and observe the PHP 8.0 deprecation warning being shown either on the page or in the error logs (depending on your test setup).
* Next, check out this branch.
* Reload the admin page and observe that the taxonomies are now sorted alphabetically within their groups.
* Also observe that the deprecation warning will no longer be thrown.

### Test instructions for QA when the code is in the RC
